src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp

```diff
@@ -1829,6 +1829,49 @@ void C2_MacroAssembler::float16_to_float(FloatRegister dst, Register src, Regist
   bind(stub->continuation());
 }
 
+static void float_to_float16_slow_path(C2_MacroAssembler& masm, C2GeneralStub<Register, FloatRegister, Register>& stub) {
+#define __ masm.
+  Register dst = stub.data<0>();
+  FloatRegister src = stub.data<1>();
+  Register tmp = stub.data<2>();
+  __ bind(stub.entry());
+
+  __ fmv_x_w(dst, src);
+
+  // preserve the payloads of non-canonical NaNs.
+  __ srai(dst, dst, 13);
+  // preserve the sign bit.
+  __ srai(tmp, dst, 13);
+  __ slli(tmp, tmp, 10);
+  __ mv(t0, 0x3ff);
+  __ orr(tmp, tmp, t0);
+
+  // get the result by merging sign bit and payloads of preserved non-canonical NaNs.
+  __ andr(dst, dst, tmp);
+
+  __ j(stub.continuation());
+#undef __
+}
+
+// j.l.Float.floatToFloat16
+void C2_MacroAssembler::float_to_float16(Register dst, FloatRegister src, FloatRegister ftmp, Register xtmp) {
+  auto stub = C2CodeStub::make<Register, FloatRegister, Register>(dst, src, xtmp, 130, float_to_float16_slow_path);
+
+  // in riscv, NaN needs a special process as fcvt does not work in that case.
+
+  // check whether it's a NaN.
+  fclass_s(t0, src);
```

Comment:
As showed roundD intrinsic PR, ( https://github.com/openjdk/jdk/pull/16382/files#diff-7a5c3ed05b6f3f06ed1c59f5fc2a14ec566a6a5bd1d09606115767daa99115bdR4252 ) the feq_s(t0, src, src) + beqz(t0, label) seems to be a faster check for NaN, could you check the numbers with feq_s ?

Reply:
Thanks for the suggestion. Yes, it bring better performance.

After: 
```
Benchmark                                     (size)  Mode  Cnt     Score    Error  Units
Fp16ConversionBenchmark.floatToFloat16          2048  avgt    5  3753.179 ? 43.557  ns/op
Fp16ConversionBenchmark.floatToFloat16Memory    2048  avgt    5    19.772 ?  0.860  ns/op
```
Before:
```
Benchmark                                     (size)  Mode  Cnt     Score    Error  Units
Fp16ConversionBenchmark.floatToFloat16          2048  avgt    5  4099.820 ? 57.671  ns/op
Fp16ConversionBenchmark.floatToFloat16Memory    2048  avgt    5    20.181 ?  0.108  ns/op
```